### PR TITLE
Checking the agent status in wazuh-manager when it is closed gracefullyv2

### PR DESCRIPTION
**Pull request by pro-akim**
|Related issue|
|-------------|
| #3709 |

## Description

Wazuh manager should receive correctly the agent closing message.  The closing process will happen gracefully, therefore, this test is designed to check the status of the agent in wazuh-manager (agent_control) after the shutdown. 

### Added

-test_shutdown_message

---

| Tester             | Test path | Jenkins | Local  | OS | Commit | Notes                |
|--------------------|-----------|---------|--------|-----|--------|----------------------|
| @pro-akim  (Developer)  |   tests/system/test_shutdown_message/test_shutdown_message.py   | ⚫⚫⚫ | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/11949155/report1.zip)[:green_circle:](https://github.com/wazuh/wazuh-qa/files/11949163/report2.zip)[:green_circle:](https://github.com/wazuh/wazuh-qa/files/11949164/report3.zip) | Debian     |   https://github.com/wazuh/wazuh-qa/pull/4146/commits/502f4d58c644c88b13d6d9e3d03c1609478e9c86      | Nothing to highlight  |
| @user (Reviewer)   |           | ⚫⚫⚫ | ⚫⚫⚫  |        |         | Nothing to highlight |

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,10 +2,14 @@
...
 All notable changes to this project will be documented in this file.
 
+
 ## [4.6.0] - TBD
 
 Wazuh commit: TBD \
 Release report: TBD
+## Added
+
+- Added tests for checking agent status upon ungraceful closure.([#3709](https://github.com/wazuh/wazuh-qa/pull/3709)) \- (Tests)
```
Comment (juliamagan) [inline CHANGELOG.md:12]:
```suggestion
- Added tests for checking agent status upon ungraceful closure. ([#4146](https://github.com/wazuh/wazuh-qa/pull/4146)) \- (Tests)
```

File: tests/system/test_shutdown_message/test_shutdown_message.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        agents.append(f'wazuh-agent{number_agent+1}')
+
+    restart_cluster(testinfra_hosts + agents, host_manager)
+
+    time.sleep(T_1)
+
+    yield
+
+    for agent in agents:
+        host_manager.run_command(agent, f'{WAZUH_PATH}/bin/wazuh-control restart')
```
Comment (juliamagan) [inline tests/system/test_shutdown_message/test_shutdown_message.py:60]: Why don't we use `restart_cluster()` as before?

File: tests/system/test_shutdown_message/test_shutdown_message.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+agents = []
+number_agents = 40
+
+pytestmark = [pytest.mark.cluster, pytest.mark.big_cluster_40_agents_env]
+
+
+@pytest.fixture()
+def restart_all_agents():
+    for number_agent in range(number_agents):
+        agents.append(f'wazuh-agent{number_agent+1}')
```
Comment (juliamagan) [inline tests/system/test_shutdown_message/test_shutdown_message.py:51]: We should do this outside the fixture.

File: tests/system/test_shutdown_message/test_shutdown_message.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        assertions:
+            - Verify that all agents status became 'Disconnected' after gracefully shutdown.
+
+        input_description: Different use cases are found in the test module and include parameters.
+
+        expected_output:
+            - Gracefully closed, it is expected to find agents 'Disconected' in agent-manager
+    '''
+
+    host_manager.get_host(testinfra_hosts[0]).ansible('command', f'service wazuh-manager restart', check=False)
```
Comment (juliamagan) [inline tests/system/test_shutdown_message/test_shutdown_message.py:89]: Why do we restart again?

File: tests/system/test_shutdown_message/test_shutdown_message.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+@pytest.fixture()
+def stop_gracefully_all_agents():
+    for agent in agents:
+        host_manager.run_command(agent, f'{WAZUH_PATH}/bin/wazuh-control stop')
+
+
+def test_shut_down_message_gracefully_stopped_agent(restart_all_agents, stop_gracefully_all_agents):
+    '''
+        description: Checking shutdown message when socket is closed.
+        wazuh_min_version: 4.5.0
```
Comment (juliamagan) [inline tests/system/test_shutdown_message/test_shutdown_message.py:72]: Wazuh min version is 4.5.0 or 4.6.0?

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but some changes are required. In addition, this PR includes changes that have not been used in the test, so they may be from another development.

File: tests/system/test_shutdown_message/test_shutdown_message.py (hunk 5 above)
Comment (pro-akim) [reply tests/system/test_shutdown_message/test_shutdown_message.py]: Changed in https://github.com/wazuh/wazuh-qa/pull/4146/commits/d0592899d807833147849678e691e3da5974918c

File: tests/system/test_shutdown_message/test_shutdown_message.py (hunk 4 above)
Comment (pro-akim) [reply tests/system/test_shutdown_message/test_shutdown_message.py]: Removed in https://github.com/wazuh/wazuh-qa/pull/4146/commits/d0592899d807833147849678e691e3da5974918c

File: tests/system/test_shutdown_message/test_shutdown_message.py (hunk 3 above)
Comment (pro-akim) [reply tests/system/test_shutdown_message/test_shutdown_message.py]: Changed in https://github.com/wazuh/wazuh-qa/pull/4146/commits/d0592899d807833147849678e691e3da5974918c

File: tests/system/test_shutdown_message/test_shutdown_message.py (hunk 2 above)
Comment (pro-akim) [reply tests/system/test_shutdown_message/test_shutdown_message.py]: Changed in https://github.com/wazuh/wazuh-qa/pull/4146/commits/d0592899d807833147849678e691e3da5974918c

File: CHANGELOG.md (hunk 1 above)
Comment (pro-akim) [reply CHANGELOG.md]: Changed in https://github.com/wazuh/wazuh-qa/pull/4146/commits/d0592899d807833147849678e691e3da5974918c

**Review by juliamagan — CHANGES_REQUESTED** [review]
We need to test the changes again

**Comment by pro-akim** [comment]
## 04/07/2023

Tests were run yesterday.
Reports updated and tables filled with new data.

**Review by juliamagan — DISMISSED** [review]
LGTM

**Review by damarisg — COMMENTED** [review]
I am not an expert in our systems testing. 
I see that the issue of documenting varies a lot. I believe that these are things that we will improve from now on once they are defined.

**Review by damarisg — DISMISSED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**